jax/_src/scipy/stats/bootstrap.py

```diff
@@ -0,0 +1,170 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import NamedTuple
+
+import scipy.stats as osp_stats
+
+import jax
+from jax import lax, random, vmap
+import jax.numpy as jnp
+from jax._src.scipy.special import ndtr as ndtr, ndtri as ndtri
+from jax._src.numpy.util import _wraps
+
+
+class ConfidenceInterval(NamedTuple):
+    ci_l: jnp.ndarray
+    ci_u: jnp.ndarray
+
+
+class BootstrapResult(NamedTuple):
+    confidence_interval: ConfidenceInterval
+    standard_error: jnp.ndarray
+
+
+def _bootstrap_resample_and_compute_statistic(sample, statistic, n_resamples, rng):
+    """
+    differences against scipy's _bootstrap_resample:
+    1. last arg `rng` is jax.random.PRNGKey (v.s. int `random_state`)
+    2. `batch` arg is not needed because lax.scan auto parallelize the resampling
+    3. to save memory, it computes statistic right after resampling
+    """
+    n = sample[0].shape[-1]
+
+    def _resample_and_compute_once(rng, x):
+        idxs = random.randint(rng, shape=(n,), minval=0, maxval=n)
+        # `sample` is a tuple of sample sets, we need to apply same indexing on each sample set
+        resample = jax.tree_map(lambda data: data[..., idxs], sample)
+        next_rng = jax.random.split(rng, 1)[0]
+        return next_rng, statistic(*resample)
+
+    # xs is dummy simply for the sake of carrying loops
+    _, theta_hat_b = lax.scan(_resample_and_compute_once, rng, jnp.ones(n))
+    return theta_hat_b
+
+
+def _jackknife_resample_and_compute_statistic(sample, statistic):
+    """
+    differences against scipy's _jackknife_resample:
+    1. `batch` arg is not needed because lax.scan auto parallelize the resampling
+    2. to save memory, it computes statistic right after resampling
+    """
+    n = sample.shape[-1]
+    # these assume arg `sample` has > 1 entries
+    idxs = jnp.arange(n - 1)
+    miss_first_sample = sample[1:]
+    miss_last_sample = sample[:-1]
+
+    def _jackknife_resample_and_compute(idx, resample):
+        resample = jnp.where(idxs >= idx, miss_first_sample, miss_last_sample)
+        return idx + 1, statistic(resample)
+
+    # TODO: check if it can handle `statistic` that return multiple scalars
+    _, theta_hat_i = lax.scan(_jackknife_resample_and_compute, 0, jnp.ones(n))
```

Comment:
for `_jackknife_resample_and_compute_statistic`, I weirdly find that `vmap` runs SLOWER than `scan`.
Here is the colab notebook that I did my benchmarking:
https://colab.research.google.com/drive/1abKv-zyI-CZ4BEtXc3oh5Ebx_rcFq0dp?usp=sharing

I am not sure if its because my implementation is inefficient, or I have made some errrors.

Reply:
Interesting! On CPU `scan` and friends are generally not too bad, but I suspect if you benchmarked on GPU/TPU you'd find `scan` to be much slower than `vmap`.